vernac/vernacprop.ml

```diff
@@ -13,48 +13,27 @@
 
 open Vernacexpr
 
-let rec under_control v = v |> CAst.with_val (function
-  | VernacExpr (_, c) -> c
-  | VernacRedirect (_,c)
-  | VernacTime (_,c)
-  | VernacFail c
-  | VernacTimeout (_,c) -> under_control c
-  )
-
-let rec has_Fail v = v |> CAst.with_val (function
-  | VernacExpr _ -> false
-  | VernacRedirect (_,c)
-  | VernacTime (_,c)
-  | VernacTimeout (_,c) -> has_Fail c
-  | VernacFail _ -> true)
+(* previous semantics of this function were strange, new one is not much better? *)
+let has_Fail { CAst.v } = List.mem ControlFail v.control
```

Comment:
strange in what way?

Reply:
In the sense that `Fail Fail` has a "Fail`, but indeed maybe the comment is confusing, will remove.